### PR TITLE
KRUSTY Multiphysics Model

### PR DESCRIPTION
**Pull request by aabdelhameedd**
This PR is related to issue #394

### CODE REVIEW

**Review by GiudGiud — CHANGES_REQUESTED** [review]
thanks for the contribution. I foresee this as one of the most influential contributions to the VTB this year given the importance of KRUSTY. I think the formatting will need to be impeccable, I left some directions to use MooseDocs better. Please apply them throughout every page of the documentation

**Comment by miaoyinb** [comment]
@GiudGiud Thanks for the review. I think we resolved all the comments. It should be ready for another round of review.

I added some syntax check tests and mesh gen tests. Not sure if we want regression tests as these are kind of heavy.

**Comment by GiudGiud** [comment]
can you run a single timestep for the BISON case? 
and what about an eigenvalue calculation without multiphysics for the Griffin one?
how expensive would each of those be?
we have access to a beefier machine for Griffin now, within reason we can make some larger cases work

**Comment by ycao-anl** [comment]
The Griffin neutronic alone calculation on sawtooth is roughly about 12 minutes with 640 CPUs if using NA=3, SN35.
For NA=1 and SN13, the run time can be 20 times faster.
Yan

**Comment by GiudGiud** [comment]
> For NA=1 and SN13, the run time can be 20 times faster.

that seems workable. The tests can modify the input through cli_args so we can afford to run that for regression

**Comment by miaoyinb** [comment]
Regression cases added.

**Comment by miaoyinb** [comment]
@GiudGiud just a friendly reminder that this PR is ready for another round of review. Thanks.

**Comment by GiudGiud** [comment]
Hello!

I'll get to it once our giant submodule update PR makes it through.
Thansk for reaching it out

**Comment by miaoyinb** [comment]
@GiudGiud Just a friendly reminder that this is still pending review. Thanks in advance.

**Comment by GiudGiud** [comment]
will do! thanks for the reminder.

File: doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+Finally, [axial_output] shows the steps to tally the axial power distribution in the KRUSTY innermost (“r1”) layer. To integrate the fission power within each fuel layer (seven radial layers), an AuxVariable in MOOSE that can be processed by the VectorPostprocessor was first defined to hold the layer-integrated total fission power for the axial regions within that layer as shown in the first block of [axial_output]. A UserObject as shown in the second block of [axial_output] was necessary to perform the actual integration among the different axial layers. The integrated values were then placed into the AuxVariable as shown in the third block of [axial_output] and were printed out by the VectorPostprocessor as shown in the fourth block of [axial_output].
```
Comment (GiudGiud) [inline doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md:124]:
```suggestion
Finally, [axial_output] shows the steps to tally the axial power distribution in the KRUSTY innermost (“r1”) layer. To integrate the fission power within each fuel layer (seven radial layers), an `AuxVariable` in MOOSE that can be processed by the `VectorPostprocessor` was first defined to hold the layer-integrated total fission power for the axial regions within that layer as shown in the first block of [axial_output]. A `UserObject` as shown in the second block of [axial_output] was necessary to perform the actual integration among the different axial layers. The integrated values were then placed into the `AuxVariable` as shown in the third block of [axial_output] and were printed out by the `VectorPostprocessor` as shown in the fourth block of [axial_output].
```
to be consistent with the beginning of the text

File: doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+For a thermo-mechanical model, thermophysical properties (i.e., thermal conductivity, density, specific heat, and thermal expansion coefficient) and mechanical properties (i.e., elastic modulus) of all the involved components are needed. [thm_model] lists the BISON/MOOSE models used for thermo-mechanical simulations. BISON doesn't include models for KRUSTY’s U-7.65Mo fuel material, so the thermal conductivity of U-10Mo and specific heat of U-8Mo are currently applied. Other thermal and mechanical attributes for U-7.65Mo and B4C were sourced from the open literature or based on certain assumptions. The displacements in the “soft materials” regions are not solved using MOOSE’s tensor mechanics solver as the solid components. Instead, simple diffusion is used for these regions to simply propagate the deformation of the solid components.
```
Comment (GiudGiud) [inline doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md:166]:
```suggestion
For a thermo-mechanical model, thermophysical properties (i.e., thermal conductivity, density, specific heat, and thermal expansion coefficient) and mechanical properties (i.e., elastic modulus) of all the involved components are needed. [thm_model] lists the BISON/MOOSE models used for thermo-mechanical simulations. BISON doesn't include models for KRUSTY’s U-7.65Mo fuel material, so the thermal conductivity of U-10Mo and specific heat of U-8Mo are currently applied. Other thermal and mechanical attributes for U-7.65Mo and B4C were sourced from the open literature or based on certain assumptions. The displacements in the “soft materials” regions are not solved using MOOSE’s solid mechanics solver as the solid components. Instead, simple diffusion is used for these regions to simply propagate the deformation of the solid components.
```

File: doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+The movement of the axial reflector to insert the reactivity was modeled by forcing a Dirichlet boundary condition on the bottom of the solid assembly that includes the axial reflector to allow BISON to calculate the consequent displacement field through solving tensor mechanics equations. Two steady-state eigenvalue calculations were included in the VTB model. These calculations were performed to confirm that an adequate amount of external reactivity was introduced for initiating the transient. The first one corresponds to the initial steady state for future transient simulation, and the second steady state corresponds to the first step of the transient after the reflector is shifted upward. In this calculation, the bias in the axial direction was set in the BISON input as shown in [mov_refl] to generate the displacement in the mesh. The displacements were passed to Griffin neutronic for calculating the total reactivity inserted into the reactor using the MultiApp coupling framework.
```
Comment (GiudGiud) [inline doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md:187]: tensor -> solid

Comment (GiudGiud) [inline doc/content/microreactors/KRUSTY/Griffin-BISON_Multiphysics_Model.md:187]:
```suggestion
The movement of the axial reflector to insert the reactivity was modeled by forcing a Dirichlet boundary condition on the bottom of the solid assembly that includes the axial reflector to allow BISON to calculate the consequent displacement field through solving solid mechanics equations. Two steady-state eigenvalue calculations were included in the VTB model. These calculations were performed to confirm that an adequate amount of external reactivity was introduced for initiating the transient. The first one corresponds to the initial steady state for future transient simulation, and the second steady state corresponds to the first step of the transient after the reflector is shifted upward. In this calculation, the bias in the axial direction was set in the BISON input as shown in [mov_refl] to generate the displacement in the mesh. The displacements were passed to Griffin neutronic for calculating the total reactivity inserted into the reactor using the MultiApp coupling framework.
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
minor things. I ll test the inputs tomorrow probably on HPC

**Comment by GiudGiud** [comment]
to reduce the picture size you must remove it from the commit that introduced it. Otherwise it does not change anything with regards to the repository size

**Comment by miaoyinb** [comment]
> to reduce the picture size you must remove it from the commit that introduced it. Otherwise it does not change anything with regards to the repository size

Will do. I am working on the other comments and will do a rebase once everything looks fine.

**Comment by GiudGiud** [comment]
looks like it does not converge on HPC for me.
We could be getting unlucky, what is the exact configuration you ran? 15 nodes * 40 procs?

alternatively we should change BISON to not error but rather to use the solutionInvalidInterface here

[log.txt](https://github.com/user-attachments/files/16662555/log.txt)

**Comment by GiudGiud** [comment]
for the input as is (not changing the azimuthal, so 3,5,3) what did you use?

**Comment by miaoyinb** [comment]
I used 2 3 3

**Comment by GiudGiud** [comment]
I imagine 2,3,3 works fine?

@aabdelhameedd what was the configuration to run the input as it is currently ?

**Comment by miaoyinb** [comment]
I think we only use 3 5 3 for neutronics only. How many nodes did you use? @ycao-anl

**Comment by ycao-anl** [comment]
I used 640 CPUs.

Yan

**Comment by GiudGiud** [comment]
how did you do 640 on sawtooth?  16 nodes with 40 cpus each?
asking because sawtooth has 48 cpus per node

**Comment by miaoyinb** [comment]
Ah, for the multiphysics simulations, the power needs to be reduced as the model is for the warm critical tests without heat pipe heat removal. I will fix it soon. We use 0.01 W for steady state as the power only rises to ~900W during reactivity insertion transient.

**Comment by ycao-anl** [comment]
Yes, I used 40 cpus and 16 nodes.

**Comment by GiudGiud** [comment]
Sounds good.
Thanks @miaoyinb we should be able to roll this out soon then! VTB testing is back on now too

**Comment by GiudGiud** [comment]
I'll take it from here @miaoyinb I was waiting for the recipes to update. I think you merged it right on time though

**Comment by GiudGiud** [comment]
Test failures are unrelated, on other modles, and being addressed in current maintenance operations

**Review by GiudGiud — APPROVED** [review]
merging once Griffin testing completes

File: microreactors/KRUSTY/tests [hunk 4]
```diff
@@ -0,0 +1,84 @@
+[Tests]
+  [syntax_check]
+    [neutronics_griffin]
+      type = RunApp
+      input = 'Neutronics/Griffin/krusty_ANL22_endf70_hybrid_SN35_NA3_coarse_CMFD.i'
+      check_input = True
+      executable_pattern = 'griffin*|blue_crab*|dire_wolf*'
+      # Cross section warning
+      allow_warnings = true
```
Comment (GiudGiud) [inline microreactors/KRUSTY/tests:9]:
```suggestion
      allow_warnings = true
      # Too slow
      method = 'devel'
```

File: microreactors/KRUSTY/tests [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      # Too slow
+      method = 'devel'
```
Comment (GiudGiud) [inline microreactors/KRUSTY/tests:11]:
```suggestion
      method = 'opt'
```

**Comment by GiudGiud** [comment]
Cardinal failure unrelated, DW recipe crashed but passed earlier

**Comment by GiudGiud** [comment]
Thanks for the contribution @aabdelhameedd @ycao-anl @miaoyinb !

**Pull request merged**